kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ChecksumReader.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.replay;
+
+import static io.delta.kernel.internal.replay.VersionStats.fromColumnarBatch;
+import static io.delta.kernel.internal.util.FileNames.checksumFile;
+import static io.delta.kernel.internal.util.FileNames.isChecksumFile;
+import static io.delta.kernel.internal.util.Utils.singletonCloseableIterator;
+
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.util.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility method to load protocol and metadata from the Delta log checksum files. */
+public class ChecksumReader {
+  private static final Logger logger = LoggerFactory.getLogger(ChecksumReader.class);
+
+  /**
+   * Load the protocol and metadata from the checksum file at the given version. If the checksum
+   * file is not found at the given version, it will try to find the latest checksum file that is
+   * created after the lower bound version or within the last 100 versions.
+   *
+   * @param engine the engine to use for reading the checksum file
+   * @param logPath the path to the Delta log
+   * @param readVersion the version to read the checksum file from
+   * @param lowerBoundOpt the lower bound version to search for the checksum file
+   * @return Optional {@link VersionStats} containing the protocol and metadata, and the version of
+   *     the checksum file. If the checksum file is not found, it will return an empty
+   */
+  public static Optional<VersionStats> getVersionStats(
+      Engine engine, Path logPath, long readVersion, Optional<Long> lowerBoundOpt) {
+
```

Comment:
add some better logging please -- i.e. what we are looking for, if we found it, and the lowerbound listfrom, etc.